src/autopas/utils/logging/FLOPLogger.cpp

```diff
@@ -0,0 +1,64 @@
+/**
+ * @file FLOPLogger.cpp
+ * @author F. Gratl & S. Newcome
+ * @date 17/04/2024
+ */
+
+#include "FLOPLogger.h"
+
+#include <string>
+
+#include "Logger.h"
+#include "autopas/utils/Timer.h"
+
+autopas::FLOPLogger::FLOPLogger(const std::string &outputSuffix) : _loggerName("FLOPLogger" + outputSuffix) {
+#ifdef AUTOPAS_LOG_FLOPS
+  _loggedEmptyFields = false;
+  const auto *fillerAfterSuffix = outputSuffix.empty() or outputSuffix.back() == '_' ? "" : "_";
+  const auto outputFileName("AutoPas_FLOPCount_" + outputSuffix + fillerAfterSuffix +
+                            autopas::utils::Timer::getDateStamp() + ".csv");
+  // Start of workaround: Because we want to use an asynchronous logger we can't quickly switch patterns for the header.
+  // Create and register a non-asychronous logger to write the header.
+  const auto headerLoggerName = _loggerName + "header";
+  auto headerLogger = spdlog::basic_logger_mt(headerLoggerName, outputFileName);
+  // set the pattern to the message only
+  headerLogger->set_pattern("%v");
+  // print csv header
+  headerLogger->info(
+      "Date,"
+      "Iteration,"
+      "Number FLOPs,"
+      "Hit Rate");
+  spdlog::drop(headerLoggerName);
+  // End of workaround
+
+  // create and register the actual logger
+  auto logger = spdlog::basic_logger_mt<spdlog::async_factory>(_loggerName, outputFileName);
+  // set pattern to provide date
+  logger->set_pattern("%Y-%m-%d %T,%v");
+#endif
+}
+
+autopas::FLOPLogger::~FLOPLogger() {
+#ifdef AUTOPAS_LOG_FLOPS
+  if (_loggedEmptyFields) {
+    AutoPasLog(INFO,
+               "FLOPLogger logged some empty fields. This might be because a Functor was used which did not provide an "
+               "implementation for getNumFLOPs or getHitRate.");
+  }
+  spdlog::drop(_loggerName);
+#endif
+}
+
+void autopas::FLOPLogger::logIteration(size_t iteration, int numFLOPs, double hitRate) {
+#ifdef AUTOPAS_LOG_FLOPS
+  // Convert negative numbers to empty strings to represent no implementation
+  const auto numFLOPsStr = numFLOPs >= 0 ? std::to_string(numFLOPs) : "";
+  const auto hitRateStr = hitRate >= 0 ? std::to_string(hitRate) : "";
+  spdlog::get(_loggerName)->info("{},{},{}", iteration, numFLOPsStr, hitRateStr);
```

Comment:
What's the reason you are logging anything at all in that case and not just skip the call to the logger?

Something like:
```suggestion
  const auto numFLOPsStr = numFLOPs >= 0 ? std::to_string(numFLOPs) : "";
  const auto hitRateStr = hitRate >= 0 ? std::to_string(hitRate) : "";
  // Only write if at least one value is not empty
  if (not numFLOPsStr.empty() or not hitRateStr.empty())) {
    spdlog::get(_loggerName)->info("{},{},{}", iteration, numFLOPsStr, hitRateStr);
  }
```

Reply:
I don't want to skip the logger because it makes it clear that the functor was still called but that no metrics were provided, as opposed to that it wasn't called. Furthermore, consider the scenario that two functors are used, one of which has one function implemented only and the other has none (niche scenario, but illustrates wider point). One functor results in a log with one empty field, implying that empty fields are outputted, and the other doesn't log anything.

Ultimately, I don't think the scenario where the user has `AUTOPAS_LOG_FLOPS=ON` but is using a functor which has no implementation is that important - I just want something which works and is not going to leave the user wondering why the FLOP log makes no sense,

Reply:
> consider the scenario that two functors are used,

For that case you should probably also write the functor name in every line or create one log file per functor.

Reply:
Yes, I mean this is something we should have anyway for the iteration logger (I think the 3 body branch had this? Not sure if it still does). But I would argue that this should be a separate PR simply because its really a separate issue which transcends this particular PR. It is also a lower priority issue relative to this PR (which is required for other PRs). But by keeping the functionality as it is, we future-proof for this particular problem combined with the other reasons for keeping this functionality.

Reply:
Fine with me